contract/src/main/java/io/pravega/schemaregistry/contract/data/SerializationFormat.java

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.schemaregistry.contract.data;
+
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.Setter;
+
+/**
+ * Different types of serialization formats used for serializing data. 
+ * Registry supports Avro, Protobuf and Json serialization formats but any custom type could be used with the registry using custom type. 
+ *
+ * If a serialization format is not present in the enum it can be specified using {@link SerializationFormat#custom} with {@link SerializationFormat#customTypeName}. 
+ * Allowed values of {@link Compatibility} mode with custom type are AllowAny or DenyAll.
+ */
+
+public enum SerializationFormat {
+    Avro,
+    Protobuf,
+    Json,
+    Any,
+    Custom;
+
+    @Getter
+    @Setter(AccessLevel.PRIVATE)
+    private String customTypeName;
+
+    /**
+     * Method to define a custom serialization format with a custom name. 
+     * @param customTypeName Custom type name. 
+     * @return {@link SerializationFormat#Custom} with supplied custom type name. 
+     */
+    public static SerializationFormat custom(String customTypeName) {
```

Comment:
tkaitchuck 5 days ago Member
```
This list should also be a string. Because we don't want to create a migration problem if users start using one format and label it as Custom, only to have to migrate to an enum value here when we add it in the future.```
@shiveshr
shiveshr 5 days ago Author Collaborator
```
This is not a string but an enum because the service understands and parses the schemas for known types and these known types are expressed in the enum. Custom name is provided for supplying any other name.

    Because we don't want to create a migration problem if users start using one format and label it as Custom, only to have to migrate to an enum value here when we add it in the future.

That is a fair point. But the migration problem will still be there as the name the user chooses for the custom format may or maynot be same as the name we introduce in the service. For example, we use Protobuf but had it not been as part of enum, some people could have called it protobuf while others could have passed Protocol-buffers.
```
@tkaitchuck
tkaitchuck 3 days ago Member
```
Well, if your goal is to create a standardized list, you can still do that with static values.
If we are going to attempt to make an exhaustive list it should at least include these:
https://en.wikipedia.org/wiki/Comparison_of_data-serialization_formats
as well as mark up languages like html, css, etc, as well as formats for tables like SQL etc.

In particular I don't see CBOR or Bincode, which we will need if we want our existing code to use this.

I am not sure what to do about formats that contain other fromats, like sequence files.```
@shiveshr
shiveshr 3 days ago Author Collaborator
```
Since registry allows users to store schemas from any format so whatever list we come up with will not necessarily be exhaustive.
I can include all these.

We could have kept the SerializationFormat as a simple String name that is generic to allow for any name.
However, I wanted to standardize the naming representation of avro protobuf and json by including them in the enum string.
Reasons being:
service has parsers to parse and do at least basic validation (well formed) for these three formats. And we will support compatibility checkers for protobuf and json in near future too.
Also, service is able to present these schemas as json strings (helpful for the UI visualization).
So by standardizing on the exact "names" for these formats serves the above purpose.
We can standardize on other serialization format names too and include them in the list, but they will not get the same treatment as these three formats, at least in initial releases of the service.
```

Reply:
If there is a difference between how they are handled there should probably be a difference in how they are represented. Perhaps there should be a boolean to indicate they are validated (or expected to be validated by the client). That way a client can provide avro and have it not validated, and we can have unvalidated cbor and then turn on validation later. Rather than suddenly breaking because the server code changed and our stream wasn't following the rules as defined by the server unbeknownst to us.

Reply:
Firstly, validation is performed based on the user specified "policy". For formats that the service does not inherently understand, the policy choice of `AllowAny` and `DenyAll` are the only valid options. 
1. a user will not experience any change in behaviour because service starts supporting a new format. 
2. Whenever server code changes, it will change in backward compatible manner. So a user who has used a custom format will continue to do so. 
And when service includes fisrt class format support for a new serialization format, the service will provide a fixed name for it. That would not affect any of the existing users in any way.

Reply:
In that case the Policy should specify which formats it applies to. In which case the format here is irrelevant, it's totally arbitrary as long as matches what is stated in the policy. In which case, why not a free form string?

Reply:
Like i said, we can make it free form string. But we created an enum with a mechanism to supply the free form string where the enum lists the 3 format names we know about. So that  people know exactly how to express the three formats service knows about. While they can still use any free form string for any format they want. 

The only thing is, service imposes restriction on policy choice for free form string. 

You see, policy is a different dimension to the format. Policy tells the service what it needs to do when a new schema is added. 
Format tells the service the format of the schema. 
Now think of a `format to policy` matrix. Where the three known formats have certain policy choices as accepted values. 
For anything else, if you try to set the policy value as anything other than AllowAny or DenyAll the service prohibits it.

Reply:
>Now think of a format to policy matrix. Where the three known formats have certain policy choices as accepted values.
For anything else, if you try to set the policy value as anything other than AllowAny or DenyAll the service prohibits it.

This is really inconsistent. If we are going to ignore the compatibility policy for some format, it should be clear from the API.

I see four ways to achieve this: 
1. Have compatibility rules expressed as a generic "rule" which take as an argument what types they are applied to.
2. Have compatibility rules applied to all formats, and setting an unsupported format is implicitly "deny". (This might be undesirable)
3. Have fromats registered to a group and have policies set on a per-format basis.
4. Allow only one format per group. Users that want multiple formats can create multiple groups. Each group can have its own policy set (assuming such a policy is supported).

Feel free to choose any of these approaches.

Reply:
> If we are going to ignore the compatibility policy for some format, it should be clear from the API.

We are not ignoring - we are rejecting policy that we cannot support. 

IMO None of the 4 suggestions are cleaner as they conflate policy and format.